### ISSUE TITLE
Mobile App UserId generation using Facebook Credentials makes no sense

### ISSUE THREAD

**Issue by corneliu-serediuc** [Azure/azure-mobile-apps-net-server#82]
Hi!

I have been playing with the Facebook authentication and I have realized that the sid:nameidentifier is generated using the Facebook primary email address. If I change my primary email address in Facebook, than a new sid:nameidentifier is generated. Thus, in my mobile app I have to create another user (considering that my primary key for users is sid:nameidentifier).

Wouldn't it make more sense to generate the sid:nameidentifier from the Facebook App Scoped User Id? This value does not change if the primary email is changed in Facebook.

The initial assumption was that people use the same email address for Google, Facebook, etc. ?

Just trying to understand.

BR // Corneliu

**Comment by brettsam** [comment]
Your observations are correct. The sid is generated based on the email registered with the provider, which may change if that registered email is changed. We're discussing alternatives to this and it may change in the future (with an opt-in to any new behavior).

As it is today, I do not recommend using the sid as your unique identifier for a user. Instead, use the provider's scoped user id. You can get this by calling `GetAppServiceIdenityAsync` and checking the `Claims[ClaimTypes.NameIdentifier]` value.

I've created a new wiki topic so I can gather these questions and answers around user ids -- you can find it here: https://github.com/Azure/azure-mobile-apps-net-server/wiki/Understanding-User-Ids

Here's a snippet for how you could generate new unique user ids from within a controller (I'm sticking this in the wiki as well). The below code generates ids like "Facebook:{uniqueid}" or "Twitter:{uniqueId}". I'll investigate wrapping this all into a helper that we can include in the SDK as well. But for now it may help:

```
[Authorize]
// GET api/values
public async Task<string> Get()
{
    ClaimsPrincipal principal = this.User as ClaimsPrincipal;
    string provider = principal.FindFirst("http://schemas.microsoft.com/identity/claims/identityprovider").Value;

    ProviderCredentials creds = null;
    if (string.Equals(provider, "facebook", StringComparison.OrdinalIgnoreCase))
    {
        creds = await this.User.GetAppServiceIdentityAsync<FacebookCredentials>(this.Request);
    }
    else if (string.Equals(provider, "google", StringComparison.OrdinalIgnoreCase))
    {
        creds = await this.User.GetAppServiceIdentityAsync<GoogleCredentials>(this.Request);
    }
    else if (string.Equals(provider, "twitter", StringComparison.OrdinalIgnoreCase))
    {
        creds = await this.User.GetAppServiceIdentityAsync<TwitterCredentials>(this.Request);
    }
   /// any other providers you support in your app.

    string userName = string.Format("{0}:{1}", creds.Provider, creds.Claims[ClaimTypes.NameIdentifier]);
    return userName;
}
```

**Comment by corneliu-serediuc** [comment]
Hi @brettsam,

Thank you for the quick response. It's funny you mention this approach, because this is exactly what I have been using so far (providerName:providerId pair). What made me re-evaluate everything was this new approach that the Mobile Apps took. Ill try to be brief.

My newly migrated Mobile App from Mobile Services includes Push Notifications (just APNS for now) and SignalR. 

**1. SignalR**
When you use this beautiful function `GetAppServiceIdentityAsync` within the Api or Table controller, everything runs smooth and async. But when you try to do the same thing within a class that implements `IUserIdProvider` from SignalR by overriding the non-asyncable function `string GetUserId(IRequest request)`, things get pretty weird. And btw, this is needed if you are trying to notify a specific user through the SignalR hub (i.e `Clients.User(receiverId).addMessage(message)`). 

The part that I could not figure out was converting an `IRequest` object to a `HttpRequestMessage` object that is needed as parameter for the `GetAppServiceIdentityAsync` function. Maybe I'm missing something obvious here, but what I ended up doing was to copy a huge chunk of code from the [azure-mobile-apps repository](https://github.com/Azure/azure-mobile-apps-net-server/blob/master/src/Microsoft.Azure.Mobile.Server.Authentication/Extensions/IPrincipalExtensions.cs), because apparently the ONLY information need from all that huge `HttpRequestMessage` object is the `x-zumo-auth` token, which I already had access to within the SignalR hub through `Context.Request.Headers["x-zumo-auth"]` (my hub is also secured).

So lets say that this part was (eventually) resolved I could extract the unique providerName:providerId that I needed by deserializing the `x-zumo-auth` token.

**2. Push Notifications**
I moved to task number 2: sending push notifications to specific users. And the way I did that before was by using the wonderful unique pair (providerName:providerId) as Tag for my Apple Notification Payload. This way I made sure that the notifications arrive to the right user, not matter how many devices that user uses.

To my surprise, after reading this [fresh new article](http://blogs.msdn.com/b/writingdata_services/archive/2016/01/22/adding-push-notification-tags-from-an-azure-mobile-apps-client.aspx), I found out that you can not add custom tags anymore and the only provided tags "by design" are `_UserId:sid:nameidentifier` and `$InstallationId:{guid}` which of course make sense to me. But again, the `sid:nameidentifier` is used.

Bottom line? The common denominator for all these services is the `sid:nameidentifier` and it is really easy to retrieve it/work with it both on the server and the client when you try to integrate all these different services.

On a personal note, I think I understand your challenge. It makes more sense to generate the `sid:nameidentifier` from an email than from a providerName:providerId key, because as soon as the provider is changed, the `sid:nameidentifier` is definitely changed. Going with the email, there might still be a change that the same address is used in multiple providers.

Anyway... for now I will go with `sid:nameidentifier` as the unique key for the user. I don't want to go through the hustle of managing one-to-many relationship between user and providers. If someone changes their primary email address in Facebook, a new user will be created in my app. Lets hope the "design" of Mobile App Authentication will not change drastically in the immediate future.

BR // Corneliu

PS: sorry for the long message. I needed to get it all out there. :P

**Comment by pvyhlas** [comment]
Hello,
I experienced some strange behavior,
I use my MS account to log in. I am pretty sure that I didn't change my email.
Today I got generated different `sid:` for this same account.

How to obtain unique user identifier on client? if  `Client.CurrentUser.UserId` is no longer valid identifier.

Thanks,
PV

**Comment by brettsam** [comment]
@pvyhlas:

I see you created issue #85 already for the changing SID. I'll follow up there for that issue.

As for your other question: the code I've listed above (and on the wiki (here)[https://github.com/Azure/azure-mobile-apps-net-server/wiki/Understanding-User-Ids]) should help you figure out the current user's unique id. Whenever you make a call with an authenticated user, you can use that code to retrieve their unique id, which you can then use as a lookup for your database. 

We're looking at ways to make this easier, but this should work for now. Let me know if you run into issues.

**Comment by brettsam** [comment]
@corneliu-serediuc 

First, thanks for the feedback. You're correct that mapping one user to many providers is a tough problem to solve. Basing things on the registered email helps with some aspects but has these drawbacks as well. Feedback like this helps move us in the right direction.

1. SignalR -- Great point about the overload. You're right that all we need is that token to generate the request for the additional token content. If we add an overload here, would that solve the SignalR problems you ran into? I haven't had a chance to dive in but I will add that as a new issue shortly so we can address it.

2. Push -- You're correct that the Push endpoint is relying on the sid for registering tags and doesn't really provide a way to register your own custom tags which is unfortunate given the new guidance that you shouldn't use the sid as a unique id. I have some thoughts on how we could make this a little more extensible and allow you to tell us which user id string to use. I'll file a new issue on that and update it when I have more to share. In the meantime, if you wanted to get adventurous and switch all of your UserId mapping to use the provider-unique-id instead of the sid, you could clone our repo and update the two lines in [NotificationInstallationsController.cs](https://github.com/Azure/azure-mobile-apps-net-server/blob/master/src/Microsoft.Azure.Mobile.Server.Notifications/Controllers/NotificationInstallationsController.cs) that look for NameIdentifier to instead call GetIdentitiesAsync and use the provider-unique-id.

**Comment by chadiswar** [comment]
When calling any of the await statements such as 

await this.User.GetAppServiceIdentityAsync<FacebookCredentials>(this.Request);

The statement goes into a black hole and ever returns any data.  I have successfully logged in - in the case facebook - and when stepping through the debugger attached to azure, The statement never returns.  Any suggestions?

Chad

**Comment by fileman** [comment]
I've same problem with AAD debug breakpoint show this.User as null

**Comment by brettsam** [comment]
@chadiswar -- To help debug, can you:

1. Check if this.User is null.
2. Look at the incoming X-ZUMO-AUTH token with `request.GetHeaderOrDefault("x-zumo-auth")`. Take that string and put it in the 'Encoded' text box at http://jwt.io and see what it shows in the 'Payload' box on the 'Decoded' side. You can post the contents here although you may want to hide your exact site URL -- similar to this: https://gist.github.com/anonymous/3c4382e7c1bffcbb5495.
3. If you put a try/catch around the call, is any exception ever thrown?

**Comment by chadiswar** [comment]
Brett,

First thank you for responding.  Here is what I am seeing.
1)  The this.User is not null.
2)  Here is the contents of the token.  I am not concerned with the content.  When I started having problems I spun up a new Mobile App with all of the latest nugget packages, only the google provider defined and with just a test method defined.  Same results as my working project.

`{
  "stable_sid": "sid:33bf3d49a17facce3e8b262fd1131cef",
  "sub": "sid:f32de0a8388aa519c64dab51222847d2",
  "idp": "google",
  "ver": "3",
  "iss": "https://testauthjdb.azurewebsites.net/",
  "aud": "https://testauthjdb.azurewebsites.net/",
  "exp": 1456437761,
  "nbf": 1456434162
}`

3)  I have actually wrapped the method in a try/catch block and no it never reaches the catch.  The swagger test call will eventually return a 500 error with no content but the server attached debug session never a break point after the GetAppServiceIdentityAsync call.  I have attached the full class I am using for the testauthjdb site above.  Thank you so much for you help.
`
using System.Web.Http;
using Microsoft.Azure.Mobile.Server.Config;
using System.Threading.Tasks;
using System.Security.Claims;
using Microsoft.Azure.Mobile.Server.Authentication;
using System;
using System.Security.Principal;
using System.Net.Http;

namespace TestAPI.Controllers
{
    [MobileAppController]
    public class TestController : ApiController
    {
        [HttpPost]
        [Authorize]
        [Route("api/Test/GetId")]
        public DataModels.JdbMobileApp.GetIdResponse GetId(DataModels.JdbMobileApp.GetIdRequest id)
        {
            Task<string> sss = GetProviderId();
            sss.Wait();
            return new DataModels.JdbMobileApp.GetIdResponse { Id = sss.Result };
        }

        [Authorize]
        private async Task<string> GetProviderId()
        {
            try
            {
                ClaimsPrincipal principal = this.User as ClaimsPrincipal;
                string provider = principal.FindFirst("http://schemas.microsoft.com/identity/claims/identityprovider").Value;

                string h = this.Request.GetHeaderOrDefault("x-zumo-auth");

                ProviderCredentials creds = null;
                if (string.Equals(provider, "facebook", StringComparison.OrdinalIgnoreCase))
                {
                    creds = await User.GetAppServiceIdentityAsync<FacebookCredentials>(this.Request);
                }
                else if (string.Equals(provider, "google", StringComparison.OrdinalIgnoreCase))
                {
                    var t = await User.GetAppServiceIdentityAsync<GoogleCredentials>(this.Request);
                }
                else if (string.Equals(provider, "twitter", StringComparison.OrdinalIgnoreCase))
                {
                    creds = await User.GetAppServiceIdentityAsync<TwitterCredentials>(this.Request);
                }
                /// any other providers you support in your app.

                string userName = creds.Claims[ClaimTypes.NameIdentifier];
                return userName;
            }
            catch (Exception e)
            {
                return e.Message;
            }
        }
    }
}
`

**Comment by brettsam** [comment]
I think you're hitting the ASP.NET example of deadlocking here: http://blog.stephencleary.com/2012/07/dont-block-on-async-code.html.

I was able to repro this and by going 'async all the way down', I was able to get it working. So for you the Controller method would look something like:

```
public async Task<DataModels.JdbMobileApp.GetIdResponse> GetId(DataModels.JdbMobileApp.GetIdRequest id)
{
    string sss = await GetProviderId();
    return new DataModels.JdbMobileApp.GetIdResponse { Id = sss };
}
```

**Comment by chadiswar** [comment]
Thank you so much, this is a huge help.  My rookie status with async calls is now confirmed!  Given that I need to get the provider id for each of my methods in my Mobile App, is there any ramifications to converting them to return a public async Task<returnObject>?

Again, thank you so much.

**Comment by nhwilly** [comment]
Full disclosure: I am an even bigger rookie than anyone here.  

I am authenticating from a Xamarin native client to a Mobile App using the sample code.  

In this case, the user is not null, but it is when I call the same endpoint from a browser (after properly authenticating) the user *is* null.  

I had to update the code a little bit as the `creds.Claims` says it's deprecated and so I used `creds.UserClaims` and a little different select method to get the value out of the enumerable.

Are we expecting the value to be null there if the endpoint is being called by a web page rather than a native client?  Do I have to call a different endpoint if I'm coming from a browser than a native client?  (Tell me it ain't so - but I'll do it to make this work.)

Thanks

**Comment by brettsam** [comment]
@nhwilly: Are you referring to calling the Login API from the browser? Is that's what is giving you a null user? Or calling this GetProviderId() endpoint?

**Comment by nhwilly** [comment]
Thanks for the quick response. In a remote area with no internet.  Sitting at a bar (only wifi on the island) at the moment and don't have the code in front of me.  I just know that getting a nameidentifier claim works differently depending on client.  I need a consistent provider based unique id and other claims for all my clients.  At some point I need to do claims transformation at the server as well.  Be back at it next week.  Thanks again.

**Comment by adrianhara** [comment]
Question: how can we implement doing the above only once, e.g. after the user logs in the first time? I want to filter the data returned from the TableControllers by the logged-in user so this provider unique user id is perfect for that. 

I guess ideally one would get this provider specific id once, at user login time, and reuse it for subsequent data requests so as to not get it on every data request. However, it's not clear to me how this mechanism should work:
1. After the user logs in (e.g. using facebook) is there a hook on the server-side that we can hook into so it's automatically called and we could make the call (e.g. to facebook) to retrieve the provider specific data? If not, does this mean that the client app has to call SignIn first and then, if successful, another custom API method to get this provider specific id?

2. What about storing this ID on the client-side? Ideally I'd not do it, as it's a security risk. So then if I don't store the ID client-side (and send it with each request) it means I need to store it server-side and, on every client request, use to correct ID for that request to filter the returned data. I presume this could be implemented using a map which maps authentication tokens (which are transient and could even be revoked) to the user's provider unique ID, but in order to do that I guess again there'd have to be some hook one could hook into to be notified when a token is issued and update the map. Is there something like that, or is there a better idea to implement this scenario? (like maybe adding it to the claims)

**Comment by brettsam** [comment]
The good news is that the user claims now have enough information to construct a unique, stable id without doing anything special (see my long answer here: https://github.com/Azure/azure-mobile-apps-net-server/issues/108  -- soon to be a wiki article).

I would suggest using the data that you have with the User in each request to construct a globally-unique identifier like this code below. The reason that I suggest adding the provider before the generated sid is because there is a tiny, tiny chance that UserA from Facebook and UserB from Google generate the same sid. By sticking 'facebook' or 'google' ahead of the name, that removes that issue completely:

```
private string GetUserId(IPrincipal user)
{
    ClaimsPrincipal claimsUser = (ClaimsPrincipal)user;

    string provider = claimsUser.FindFirst("http://schemas.microsoft.com/identity/claims/identityprovider").Value;
    string sid = claimsUser.FindFirst(ClaimTypes.NameIdentifier).Value;

    return $"{provider}|{sid}";
}
```

And then wherever you need to use this user id -- either for querying or for associating data items with a user, you can use this string. As an example:

```
[Authorize]
// GET tables/TodoItem
public IQueryable<TodoItem> GetAllTodoItems()
{
    string userId = GetUserId(this.User);
    return Query().Where(t => t.UserId == userId);
}

[Authorize]
// POST tables/TodoItem
public async Task<IHttpActionResult> PostTodoItem(TodoItem item)
{
    item.UserId = GetUserId(this.User);
    TodoItem current = await InsertAsync(item);
    return CreatedAtRoute("Tables", new { id = current.Id }, current);
}
```

Does that address your scenario? If not, I can dig in some more.

**Comment by nhwilly** [comment]
Thank you.  I read #108 and `stable_id` is huge help for me.

**Comment by brettsam** [comment]
Note: I've updated the wiki topic https://github.com/Azure/azure-mobile-apps-net-server/wiki/Understanding-User-Ids with details pulled from my comments on related issues. I'll continue to compile helpful snippets there.

**Comment by adrianhara** [comment]
Awesome, great and clear explanations, thanks a lot!

**Comment by adrianhara** [comment]
I updated everything as per the explanations here. Server-side it works great, with the WEBSITE_AUTH_HIDE_DEPRECATED_SID set to true.

I'm having however a problem on the client-side. After `SignInAsync()` I was getting the `MobileServiceClient.CurrentUser.MobileServiceAuthenticationToken` and decoding it with `FromBase64String()` then deserializing it and storing it locally. This doesn't seem to work anymore, specifically `MobileServiceClient.CurrentUser.MobileServiceAuthenticationToken` does not seem to be a valid base64 string, `FromBase64String()` throws.

Any idea why this might be happening?

**Comment by brettsam** [comment]
If you are trying to store it locally so you can re-use it when the app restarts, can you follow the guide here: https://azure.microsoft.com/en-us/documentation/articles/app-service-mobile-windows-store-dotnet-get-started-users/#tokens. I don't see any conversions there so it may work.

If you need the Base64 conversion, I can take a deeper look.

**Comment by nhwilly** [comment]
FWIW, I started writing the tokens to Blob Storage and noticed there's an "encrypted" flag that's set to True.  Maybe that's what's coming back.

**Comment by adrianhara** [comment]
Not sure I need the Base64 conversion if there is another way to check if the token expired locally, without hitting the server. Before I was parsing the token (from base 64 then deserializing to  a class) and looking at the "exp" property to determine if it was expired. 

Now I can't do that anymore since the base64 decoding fails. Any idea how to check the expiration client-side?
